clash-cores/src/Clash/Cores/Xilinx/DcFifo/Implementation.hs

```diff
@@ -0,0 +1,261 @@
+{-|
+Copyright  :  (C) 2022, Google Inc,
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Support for the [FIFO Generator v13.2](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator).
+
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Clash.Cores.Xilinx.DcFifo.Implementation
+  ( -- * Instantiating IP
+    dcFifo
+  , XilinxFifo(..)
+
+    -- * Customizing IP
+  , DcConfig(..)
+  , defConfig
+
+    -- * Helper type aliases
+  , Full
+  , Empty
+  , DataCount
+  ) where
+
+import           Clash.Annotations.Primitive (Primitive (InlineYamlPrimitive))
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+
+import qualified Data.Sequence               as Seq
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+-- | Parameters for the FIFO generator; toggle various signals.
+data DcConfig depth = DcConfig
+  { dcDepth          :: !depth
+  , dcReadDataCount  :: !Bool -- ^ Enable @rd_cnt@ signal
+  , dcWriteDataCount :: !Bool -- ^ Enable @wr_cnt@ signal
+  , dcOverflow       :: !Bool -- ^ Enable @overflow@ signal
+  , dcUnderflow      :: !Bool -- ^ Enable @underflow@ signal
+  }
+  deriving (Show, Generic)
+
+-- | Default config. Read\/write data counts are enabled but over-\/underflow
+-- are not.
+defConfig :: KnownNat depth => DcConfig (SNat depth)
+defConfig = DcConfig
+  { dcDepth = SNat
+  , dcReadDataCount = True
+  , dcWriteDataCount = True
+  , dcOverflow = False
+  , dcUnderflow = False
+  }
+
+data FifoState a = FifoState
+  { hsQueue      :: Seq.Seq a
+  , relativeTime :: Int -- ^ In order to model a FIFO in two clock domains,
+                        -- we track relative offset in order to know which
+                        -- signals to peel off.
+  } deriving Show
+
+data XilinxFifo read write depth a =
```

Comment:
Please also document what this datatype is (good work on documenting the fields though!).

I'd also like to bikeshed the naming. I get the sense some names aren't well suited to users importing them unqualified, like `writeCount`, `isEmpty`, `Empty`, `defConfig`. Yet others are, like `XilinxFifo` and arguably `dcFifo`. I think we need to pick a lane and be consistent, and I think the lane where users are expected to import qualified is preferred. So I think `dcFifo` is okay as the name of the function even though it is already implied that it is dual clock by the module name, but this `XilinxFifo` data type needs to be named `FifoOut` or similar, without `Xilinx`.

Reply:
Agreed 💯

Reply:
Sure, I've changed it to `FifoOut.`